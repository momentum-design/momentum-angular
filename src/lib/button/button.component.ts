/** @component button */
import {
  AfterViewInit,
  Component,
  ElementRef,
  Input,
} from '@angular/core';

export type ButtonType = 'button' | 'reset' | 'submit';

@Component({
  selector: 'button[md-button], a[md-button], input[md-button]',
  exportAs: 'mdButton',
  template: `
    <md-loading *ngIf="loading"></md-loading>
    <span class="md-button__children" [ngStyle]="{ opacity: opacity }">
      <ng-content></ng-content>
    </span>
  `,
  host: {
    '[attr.active]': 'active',
    '[attr.disabled]': 'disabled || loading || null',
    '[attr.href]': 'href',
    '[attr.tabindex]': 'index === focusIndex ? 0 : -1',
    '[attr.type]': 'type',
    'class': 'md-button md-button--36',
    '[class.active]': 'active && !disabled',
    '[class.md-button--circle]': 'circle',
    '[class.md-button--disabled]': 'disabled || loading',
    '[class.md-button--expand]': 'expand',
    '[class.md-button--none]': 'removeStyle',
  }
})
export class ButtonComponent implements AfterViewInit {
  /** @option Active Sets button to active | false */
  @Input() active: boolean = false;
  /** @option Creates a circle shaped button | false */
  @Input() circle: boolean = false;
  /** @option Sets optional button color | '' */
  @Input()
  get color(): string {
    if (this.removeStyle) {
      this.color = null;
      return null;
    }
    return this._color;
  }
  set color(value: string) {
    const colorName = (name: string) => name === 'none' ? 'color-none' : name;
    if (this._color) {
      this.el.nativeElement.classList.remove(`md-button--${colorName(this._color)}`);
    }
    if (value && !this.removeStyle) {
      this.el.nativeElement.classList.add(`md-button--${colorName(value)}`);
    }
    this._color = value;
  }
  /** @option Sets the attribute disabled to Button | false */
  @Input() disabled: boolean = false;
  /** @option Sets expand css styling to widen the Button | false */
  @Input() expand: boolean = false;
  /** @option Sets href for an anchor tagged button | '' */
  @Input() href: string = '';
  /** @option Activates the loading animation and sets the button to disabled | false */
  @Input() loading: boolean = false;
  /** @option Boolean to remove Button's default style | false */
  @Input()
  get removeStyle(): boolean {
    return this._removeStyle;
  }
  set removeStyle(value: boolean) {
    if (value) {
      this.color = null;
      this.size = null;
    }
    this._removeStyle = value;
  }
  /** @option Size value | '36' */
  @Input()
  get size(): number | string {
    if (this.removeStyle) {
      this.size = null;
      return null;
    }
    return this._size;
  }
  set size(value: number | string) {
    const sizeName = (name: number | string) => name === 'none' ? 'size-none' : name;
    if (this._size) {
      this.el.nativeElement.classList.remove(`md-button--${sizeName(this._size)}`);
    }
    this._size = value;
    if (!this.circle && !this.checkButtonSize(value)) {
      console.warn(
        `[@collab-ui/angular] Button: selected size is not supported
         for non-circular button. Size will default to 36`
      );
      this._size = 36;
    }
    if (this._size && !this.removeStyle) {
      this.el.nativeElement.classList.add(`md-button--${sizeName(this._size)}`);
    }
  }
  /** @option Html button type | 'button' */
  @Input() type: ButtonType = 'button';

  private _color: string = '';
  private _focusIndex: number;
  private _prevFocusIndex: number;
  private _removeStyle: boolean = false;
  private _size: number | string = 36;
  private _width: string;
  public focusOnLoad: boolean;
  public index: number;

  get focusIndex(): number {
    return this._focusIndex;
  }
  set focusIndex(value: number) {
    this._focusIndex = value;
    if (this.index === this._focusIndex && this._focusIndex !== this._prevFocusIndex) {
      this.el.nativeElement.focus();
    }
    this._prevFocusIndex = this._focusIndex;
  }
  get hostElement(): ElementRef {
    return this.el;
  }
  get opacity(): number {
    return this.loading ? 0 : 1;
  }
  get width(): string {
    return this._width;
  }
  set width(value: string) {
    this._width = value;
    this.el.nativeElement.style.width = value;
  }

  constructor(private el: ElementRef) {}

  ngAfterViewInit() {
    setTimeout(() => {
      this.checkAriaLabel();

      if (this.focusOnLoad && this.focusIndex === this.index) {
        this.el.nativeElement.focus();
      }
    }, 0);
  }

  private checkButtonSize(size: number | string): boolean {
    return [null, 'none', '28', '36', '40', '52', 28, 36, 40, 52].includes(size);
  }

  private checkAriaLabel() {
    const ariaLabel = this.getAttr('aria-label');
    if (!ariaLabel && this.el.nativeElement.children.length > 0) {
      throw new Error(
        'md-button: content is not a string, you must add an "ariaLabel" for accessibility.'
      );
    }
  }

  private getAttr(attr: string) {
    return this.el.nativeElement.getAttribute(attr);
  }
}
