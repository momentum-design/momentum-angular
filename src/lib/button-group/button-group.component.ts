import {
  AfterContentInit,
  Component,
  ContentChildren,
  ElementRef,
  EventEmitter,
  Input,
  Output,
  QueryList,
  Renderer2,
} from '@angular/core';
import { ButtonComponent } from '../button/button.component';

export type ButtonGroupTheme = '' | 'dark';
export type ButtonGroupType = '' | 'pill';

export interface ButtonGroupSelectChange {
  event: MouseEvent;
  index: number;
}

@Component({
  selector: 'md-button-group',
  template: `<ng-content></ng-content>`,
  host: {
    '[attr.aria-label]': 'ariaLabel',
    'class': 'md-button-group',
    '[class.md-button-group--justified]': 'justified',
    'role': 'group',
  },
})
export class ButtonGroupComponent implements AfterContentInit {
  /** @prop Sets initial active Button by index | null */
  @Input() activeIndex: number;
  /** @prop Text to display for blindness accessibility features | '' */
  @Input() ariaLabel: string = '';
  /** @prop Set focus to ButtonGroup when page is loaded | false */
  @Input() focusOnLoad: boolean = false;
  /** @prop Highlights the selected button within group | true */
  @Input() highlightSelected: boolean = true;
  /** @prop Optional text-justified css styling | true */
  @Input() justified: boolean = true;
  /** @prop Sets width of a pill Button | '60px' */
  @Input() pillWidth: string = '60px';
  /** @prop Optional Button color theme for ButtonGroup | '' */
  @Input()
  get theme(): ButtonGroupTheme {
    return this._theme;
  }
  set theme(value: ButtonGroupTheme) {
    this.elementRef.nativeElement.classList.remove(`md-button-group--${this._theme}`);
    this.elementRef.nativeElement.classList.add(`md-button-group--${value}`);
    this._theme = value;
  }
  /** @prop Optional Button type for ButtonGroup | '' */
  @Input()
  get type(): ButtonGroupType {
    return this._type;
  }
  set type(value: ButtonGroupType) {
    this.elementRef.nativeElement.classList.remove(`md-button-group--${this._type}`);
    this.elementRef.nativeElement.classList.add(`md-button-group--${value}`);
    this._type = value;
  }

  @Output() readonly select: EventEmitter<ButtonGroupSelectChange> = new EventEmitter<ButtonGroupSelectChange>();

  @ContentChildren(ButtonComponent) buttonList: QueryList<ButtonComponent>;

  private _focusIndex: number = 0;
  private _theme: ButtonGroupTheme = '';
  private _type: ButtonGroupType = '';

  constructor(private elementRef: ElementRef, private renderer: Renderer2) {}

  ngAfterContentInit() {
    this.buttonList.forEach((button, idx) => {
      button.focusOnLoad = this.focusOnLoad;
      button.index = idx;
      if (this.pillWidth) {
        button.width = this.pillWidth;
      }
      this.renderer.listen(
        button.hostElement.nativeElement,
        'click',
        e => this.handleClick(e, button.index)
      );
      this.renderer.listen(
        button.hostElement.nativeElement,
        'keydown',
        e => this.handleKeydown(e, button.index)
      );
    });

    const initialFocus = this.getNewIndex(this._focusIndex - 1 , 1);
    this.setFocusIndex(initialFocus);
    if (this.activeIndex !== null) {
      this.determineInitialActive();
    }
  }

  getNewIndex(currentIndex: number, change: number): number {
    const length = this.buttonList.length - 1;

    const getPossibleIndex = () => {
      if (currentIndex + change < 0) {
        return length;
      } else if (currentIndex + change > length) {
        return 0;
      }
      return currentIndex + change;
    };

    const possibleIndex = getPossibleIndex();
    const potentialTarget: ButtonComponent = this.buttonList.toArray()[possibleIndex];

    return potentialTarget.disabled
      ? this.getNewIndex(possibleIndex, change)
      : possibleIndex;
  }

  setFocusIndex(index: number) {
    this._focusIndex = index;
    this.buttonList.forEach(button => {
      button.focusIndex = index;
    });
  }

  determineInitialActive() {
    if (this.activeIndex < 0 && this.activeIndex > this.buttonList.length - 1) {
      console.warn('[@collab-ui/angular] ButtonGroup: activeIndex is out of bound');
      return;
    }
    if (!isNaN(this.activeIndex)) {
      const initialActive = this.getNewIndex(this.activeIndex - 1 , 1);
      this.setActiveIndex(initialActive);
    }
  }

  setActiveIndex(index: number) {
    if (this.activeIndex !== index) {
      this.activeIndex = index;
      this.buttonList.forEach((button, idx) => {
        button.active = this.type === 'pill' ? false : this.highlightSelected && this.activeIndex === idx;
      });
    }
  }

  getIncludesFirstCharacter(str: string, char: string): boolean {
    return str
      .charAt(0)
      .toLowerCase()
      .includes(char);
  }

  setFocusByFirstCharacter(char: string, currentIdx: number) {
    const length = this.buttonList.length - 1;
    const newIndex = this.buttonList
      .toArray()
      .reduce((agg, _, idx, arr) => {
        const index = currentIdx + idx + 1 > length
          ? Math.abs(currentIdx + idx - length)
          : currentIdx + idx + 1;

        const label = arr[index].hostElement.nativeElement.innerText;

        return (
          !agg.length
          && !arr[index].disabled
          && this.getIncludesFirstCharacter(label, char)
        )
          ? agg.concat(index)
          : agg;
      },
      []
    );
    if (!isNaN(newIndex[0])) {
      this.setFocusIndex(newIndex[0]);
    }
  }

  handleClick(event: MouseEvent, index: number) {
    this.setFocusIndex(index);
    this.setActiveIndex(index);
    this.select.emit({event: event, index: index});
  }

  handleKeydown(event: KeyboardEvent, index: number) {
    let newIndex: number;
    let flag: boolean = false;
    const char: string = event.key;

    const isPrintableCharacter = (str: string) => {
      return str.length === 1 && str.match(/\S/);
    };

    switch (event.code) {
      case 'ArrowLeft':
      case 'ArrowUp':
        newIndex = this.getNewIndex(index, -1);
        this.setFocusIndex(newIndex);
        flag = true;
        break;

      case 'ArrowRight':
      case 'ArrowDown':
        newIndex = this.getNewIndex(index, 1);
        this.setFocusIndex(newIndex);
        flag = true;
        break;

      default:
        if (isPrintableCharacter(char)) {
          this.setFocusByFirstCharacter(char, index);
          flag = true;
        }
        break;
    }

    if (flag) {
      event.stopPropagation();
      event.preventDefault();
    }
  }
}
