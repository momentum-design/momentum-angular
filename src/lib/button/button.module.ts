import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ButtonComponent } from './button.component';
import { LoadingModule } from '../loading';

@NgModule({
  imports: [
    CommonModule,
    LoadingModule.forRoot(),
  ],
  exports: [ButtonComponent],
  declarations: [ButtonComponent],
})
export class ButtonModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: ButtonModule,
      providers: [],
    };
  }
}
