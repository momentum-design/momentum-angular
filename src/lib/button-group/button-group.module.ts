import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ButtonGroupComponent } from './button-group.component';

@NgModule({
  declarations: [ButtonGroupComponent],
  imports: [CommonModule],
  exports: [ButtonGroupComponent],
})
export class ButtonGroupModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: ButtonGroupModule,
      providers: [],
    };
  }
}
