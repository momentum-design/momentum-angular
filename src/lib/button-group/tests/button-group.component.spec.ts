import { Component } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { ButtonGroupModule } from '../button-group.module';
import { ButtonModule } from '../../button';

describe('ButtonGroupComponent', () => {
  let testComponent: TestAppComponent;
  let fixture: ComponentFixture<TestAppComponent>;

  @Component({
    selector: 'test-app',
    template: `
      <md-button-group class="default">
        <button md-button aria-label="1">one</button>
        <button md-button aria-label="2" [disabled]="true">two</button>
      </md-button-group>

      <md-button-group class="pill" type="pill">
        <button md-button aria-label="1">one</button>
        <button md-button aria-label="2">two</button>
      </md-button-group>

      <md-button-group class="highlightSelected" [highlightSelected]="false">
        <button md-button aria-label="1">one</button>
        <button md-button aria-label="2">two</button>
      </md-button-group>

      <md-button-group class="activeIndex" activeIndex=1>
        <button md-button aria-label="1">one</button>
        <button md-button aria-label="2">two</button>
      </md-button-group>

      <md-button-group class="focusOnLoad" [focusOnLoad]="true">
        <button md-button aria-label="1" id="focus-one">one</button>
        <button md-button aria-label="2" id="focus-two">two</button>
      </md-button-group>
    `,
  })
  class TestAppComponent {}

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [TestAppComponent],
      imports: [ButtonGroupModule, ButtonModule],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestAppComponent);
    testComponent = fixture.componentInstance;
  });

  it('should create and match snapshot', () => {
    fixture.detectChanges();
    expect(testComponent).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('ButtonGroup should have justified as true', () => {
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector('.default');
    expect(element.className).toContain('md-button-group--justified');
  });

  it('ButtonGroup type is set to pill', () => {
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector('.pill');
    expect(element.className).toContain('md-button-group--pill');
  });

  it('should not highlight the active button when highlightSelected is false', () => {
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector('.highlightSelected');
    const button = element.querySelector('button');
    button.click();
    expect(button.className).not.toContain('active');
  });

  it('onClick should should mark the button as active', () => {
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector('.default');
    const button = element.querySelector('button');
    button.click();
    fixture.detectChanges();
    expect(button.className).toContain('active');
  });

  it('if focusOnLoad is set to true, then should set the focus on the first button', () => {
    fixture.detectChanges();
    expect(document.activeElement.id).toEqual('focus-one');
  });

  it('when activeIndex prop is passed, the button should be selected', () => {
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector('.activeIndex');
    const buttons = element.querySelectorAll('button');
    expect(buttons[1].className).toContain('active');
  });

  it('on keyboard events arrow right', () => {
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector('.focusOnLoad');
    const buttons = element.querySelectorAll('button');
    buttons[0].dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'ArrowRight',
      })
    );
    fixture.detectChanges();
    expect(buttons[1].getAttribute('tabindex')).toEqual('0');
  });

  it('on keyboard events arrow down', () => {
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector('.focusOnLoad');
    const buttons = element.querySelectorAll('button');
    buttons[0].dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'ArrowDown',
      })
    );
    fixture.detectChanges();
    expect(buttons[1].getAttribute('tabindex')).toEqual('0');
  });

  it('on keyboard events arrow left', () => {
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector('.focusOnLoad');
    const buttons = element.querySelectorAll('button');
    buttons[0].dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'ArrowLeft',
      })
    );
    fixture.detectChanges();
    expect(buttons[1].getAttribute('tabindex')).toEqual('0');
  });

  it('on keyboard events arrow up', () => {
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector('.focusOnLoad');
    const buttons = element.querySelectorAll('button');
    buttons[0].dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'ArrowUp',
      })
    );
    fixture.detectChanges();
    expect(buttons[1].getAttribute('tabindex')).toEqual('0');
  });

});
