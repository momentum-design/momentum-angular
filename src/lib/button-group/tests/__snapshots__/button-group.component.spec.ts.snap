// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroupComponent should create and match snapshot 1`] = `
<test-app>
  <md-button-group
    aria-label=""
    class="default md-button-group md-button-group--justified"
    role="group"
  >
    <button
      active="false"
      aria-label="1"
      class="md-button md-button--36"
      href=""
      md-button=""
      style="width: 60px;"
      tabindex="0"
      type="button"
    >
      
      <span
        class="md-button__children"
        ng-reflect-ng-style="[object Object]"
        style="opacity: 1;"
      >
        one
      </span>
    </button>
    <button
      active="false"
      aria-label="2"
      class="md-button md-button--36 md-button--disabled"
      disabled="true"
      href=""
      md-button=""
      ng-reflect-disabled="true"
      style="width: 60px;"
      tabindex="-1"
      type="button"
    >
      
      <span
        class="md-button__children"
        ng-reflect-ng-style="[object Object]"
        style="opacity: 1;"
      >
        two
      </span>
    </button>
  </md-button-group><md-button-group
    aria-label=""
    class="pill md-button-group md-button-group--pill md-button-group--justified"
    ng-reflect-type="pill"
    role="group"
    type="pill"
  >
    <button
      active="false"
      aria-label="1"
      class="md-button md-button--36"
      href=""
      md-button=""
      style="width: 60px;"
      tabindex="0"
      type="button"
    >
      
      <span
        class="md-button__children"
        ng-reflect-ng-style="[object Object]"
        style="opacity: 1;"
      >
        one
      </span>
    </button>
    <button
      active="false"
      aria-label="2"
      class="md-button md-button--36"
      href=""
      md-button=""
      style="width: 60px;"
      tabindex="-1"
      type="button"
    >
      
      <span
        class="md-button__children"
        ng-reflect-ng-style="[object Object]"
        style="opacity: 1;"
      >
        two
      </span>
    </button>
  </md-button-group><md-button-group
    aria-label=""
    class="highlightSelected md-button-group md-button-group--justified"
    ng-reflect-highlight-selected="false"
    role="group"
  >
    <button
      active="false"
      aria-label="1"
      class="md-button md-button--36"
      href=""
      md-button=""
      style="width: 60px;"
      tabindex="0"
      type="button"
    >
      
      <span
        class="md-button__children"
        ng-reflect-ng-style="[object Object]"
        style="opacity: 1;"
      >
        one
      </span>
    </button>
    <button
      active="false"
      aria-label="2"
      class="md-button md-button--36"
      href=""
      md-button=""
      style="width: 60px;"
      tabindex="-1"
      type="button"
    >
      
      <span
        class="md-button__children"
        ng-reflect-ng-style="[object Object]"
        style="opacity: 1;"
      >
        two
      </span>
    </button>
  </md-button-group><md-button-group
    activeindex="1"
    aria-label=""
    class="activeIndex md-button-group md-button-group--justified"
    ng-reflect-active-index="1"
    role="group"
  >
    <button
      active="false"
      aria-label="1"
      class="md-button md-button--36"
      href=""
      md-button=""
      style="width: 60px;"
      tabindex="0"
      type="button"
    >
      
      <span
        class="md-button__children"
        ng-reflect-ng-style="[object Object]"
        style="opacity: 1;"
      >
        one
      </span>
    </button>
    <button
      active="true"
      aria-label="2"
      class="md-button md-button--36 active"
      href=""
      md-button=""
      style="width: 60px;"
      tabindex="-1"
      type="button"
    >
      
      <span
        class="md-button__children"
        ng-reflect-ng-style="[object Object]"
        style="opacity: 1;"
      >
        two
      </span>
    </button>
  </md-button-group><md-button-group
    aria-label=""
    class="focusOnLoad md-button-group md-button-group--justified"
    ng-reflect-focus-on-load="true"
    role="group"
  >
    <button
      active="false"
      aria-label="1"
      class="md-button md-button--36"
      href=""
      id="focus-one"
      md-button=""
      style="width: 60px;"
      tabindex="0"
      type="button"
    >
      
      <span
        class="md-button__children"
        ng-reflect-ng-style="[object Object]"
        style="opacity: 1;"
      >
        one
      </span>
    </button>
    <button
      active="false"
      aria-label="2"
      class="md-button md-button--36"
      href=""
      id="focus-two"
      md-button=""
      style="width: 60px;"
      tabindex="-1"
      type="button"
    >
      
      <span
        class="md-button__children"
        ng-reflect-ng-style="[object Object]"
        style="opacity: 1;"
      >
        two
      </span>
    </button>
  </md-button-group>
</test-app>
`;
